core/src/main/resources/error/error-classes.json

```diff
@@ -75,6 +75,23 @@
       "The value <str> (<fmt>) cannot be converted to <targetType> because it is malformed. Correct the value as per the syntax, or change its format. Use <suggestion> to tolerate malformed input and return NULL instead."
     ]
   },
+  "DATATYPE_MISMATCH" : {
+    "message" : [
+      "Cannot resolve <sqlExpr> due to data type mismatch:"
+    ],
+    "subClass" : {
+      "BINARY_OP_DIFF_TYPES" : {
+        "message" : [
+          "the left and right arguments of the binary operator have different types (<left> and <right>)."
```

Comment:
```suggestion
          "the left and right operands of the binary operator have incompatible types (<left> and <right>)."
```

Reply:
One of properties of any binary operator is to have the same types of its inputs, see
https://github.com/apache/spark/blob/fab65f4e393359b9523e34457c877083dba0e7e5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L738-L742

This is what the error says precisely. I am not sure that we should introduce the term of incompatibility here.

Reply:
"If the two inputs have different types, the analyzer will find the tightest common type and do the proper type casting."
We always need to look at errors from a user's perspective...